### PR TITLE
detect and limit package cycles

### PR DESCRIPTION
**Pull request by turadg**
## Description

https://github.com/Agoric/agoric-sdk/pull/7260 burned down some cycles in service of https://github.com/Agoric/agoric-sdk/issues/4645 but it lingered and now it's pretty stale.

I'll wait for a better time and plan before updating that which does refactorings.

Meanwhile, it's worth getting in this script that detects `dependencies` cycles (whereas Lerna conflates those with `devDependencies`). This also removes some declared dependencies that aren't actual.

### Testing Considerations

CI

### CODE REVIEW

File: scripts/graph.sh [hunk 1]
```diff
@@ -19,4 +19,15 @@ DIR=$(dirname -- "${BASH_SOURCE[0]}")
...
+} | dot -Tcanon >packages-graph.dot
+dot -Tpng <packages-graph.dot >"$DIR"/../packages-graph.png
+
+dot -Tsvg <packages-graph.dot >"$DIR"/../packages-graph.svg
+
+if acyclic packages-graph.dot | dot -Tcanon >packages-sans-cycles.dot; then
+    echo "No cycles in 'dependencies' of packages."
+else
+    echo "Cycles detected. These lines appear only in the original graph and not the acyclic variant:"
+    comm -23 <(sort packages-graph.dot) <(sort packages-sans-cycles.dot)
```
Comment (dckc) [inline scripts/graph.sh:32]: I had to look up `comm -23` and `<(...)` [process substitution syntax](https://unix.stackexchange.com/a/596893). I was about to ask for a comment, but the `echo` line above says what it's doing. Pretty clearly not posix `sh`, but it says `#!/bin/bash` on the tin. [ShellCheck](https://www.shellcheck.net/) says "No issues detected!"

For ref: `man comm` says:

```
       -2     suppress column 2 (lines unique to FILE2)

       -3     suppress column 3 (lines that appear in both files)
```

**Review by dckc — COMMENTED** [review]
I'd approve this but for compatibility risks. One would think they're negligible, but... it all seems complicated just now.

**Comment by ivanlei** [comment]
This looks good overall. Is there a doc or README or something talking about how to use the tools we have in the repo to help detect dependencies? This seems too cool to only discover via stumbling upon it or tribal knowledge.

**Comment by turadg** [comment]
> Is there a doc or README or something talking about how to use the tools we have in the repo to help detect dependencies?

No. I can't think of a doc people are likely to learn of it from either. Instead I've put it into CI, to enforce a threshold. It'll be discovered as necessary :)

**Comment by turadg** [comment]
CI failing because it doesn't have `dot`. Thoughts on whether the job is worth installing GraphViz?

**Comment by michaelfig** [comment]
> Thoughts on whether the job is worth installing GraphViz?

I think that it's worth the install.  Just do it within the Github Actions workflow (with `apt`, since our runners are Ubuntu), then the bash script can continue to assume it is present.

**Comment by turadg** [comment]
The new `dependncy-graph` job passed correctly: https://github.com/Agoric/agoric-sdk/actions/runs/5259906962/jobs/9506200446

It installs `graphviz` each time. That could be cached but it only takes 20s. The whole job took 26s.

Once this PR is approved, I'll merge this but before closing out the ticket we'll need to add it to the Required checks.

File: .github/workflows/test-all-packages.yml [hunk 2]
```diff
@@ -40,6 +40,17 @@ jobs:
...
+  dependency-graph:
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Install graphviz
+        run: sudo apt install -y graphviz
+
+      - name: Check for cycles
+        run: scripts/check-dependency-cycles.sh 4
```
Comment (dckc) [inline .github/workflows/test-all-packages.yml:52]: Do you know the 4 existing cycles off-hand? If so, jot them down in reply, please?

**Review by dckc — APPROVED** [review]
all build stuff; no runtime-impact, so I don't see any compatibility risk

**Comment by erights** [comment]
None of this is specific to agoric-sdk. We need it in Endo as well. If it were in Endo, agoric-sdk could use it, but not vice versa. How about if it is just in Endo and used by agoric-sdk?

File: scripts/graph.sh [hunk 3]
```diff
@@ -19,4 +19,15 @@ DIR=$(dirname -- "${BASH_SOURCE[0]}")
...
     '
     echo '}'
-} | dot -Tpng > "$DIR"/../packages.png
+    # normalize
+} | dot -Tcanon >packages-graph.dot
+dot -Tpng <packages-graph.dot >"$DIR"/../packages-graph.png
+
+dot -Tsvg <packages-graph.dot >"$DIR"/../packages-graph.svg
+
+if acyclic packages-graph.dot | dot -Tcanon >packages-sans-cycles.dot; then
```
Comment (kriskowal) [inline scripts/graph.sh:28]: TIL `man acyclic`

**Review by kriskowal — APPROVED** [review]

File: .github/workflows/test-all-packages.yml (hunk 2 above)
Comment (turadg) [reply .github/workflows/test-all-packages.yml:52]:
```
❯ scripts/graph.sh
Cycles detected. These lines appear only in the original graph and not the acyclic variant:
	"@agoric/governance" -> "@agoric/vats";
	"@agoric/inter-protocol" -> agoric;
	"@agoric/vats" -> "@agoric/inter-protocol";
	"@agoric/zone" -> "@agoric/vat-data";
```

**Comment by turadg** [comment]
> None of this is specific to agoric-sdk.

Correct.
> We need it in Endo as well

Ok. https://github.com/endojs/endo/pull/1633

>  If it were in Endo, agoric-sdk could use it, but not vice versa. How about if it is just in Endo and used by agoric-sdk?

These are in the repo's scripts directory and CI config, so they can't be imported between repos.

**Pull request merged**